### PR TITLE
Jwt claim mapping

### PR DESCRIPTION
**Pull request by korroz**
Makes claim mapping and filtering in `JwtSecurityTokenHandler` a feature on a per instance basis.

Summary:
 - Adds public properties on instances of `JwtSecurityTokenHandler`.
    - `InstanceInboundClaimTypeMap`
    - `InstanceOutboundClaimTypeMap`
    - `InstanceInboundClaimFilter`
 - Pulls outbound claim type mapping from the `JwtPayload` class, it's now done in the handler.
 - Test added for exercising these new features.

So, as requested in #95, here you go. I hope this is a good start that can be shaped into a final api for these features.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/208%23issuecomment-127783346%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/208%23issuecomment-127791069%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/208%23issuecomment-128459310%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/208%23issuecomment-127783346%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Hi%20__%40korroz__%2C%20I%27m%20your%20friendly%20neighborhood%20Microsoft%20Pull%20Request%20Bot%20%28You%20can%20call%20me%20MSBOT%29.%20Thanks%20for%20your%20contribution%21%5Cr%5Cn%20%20%20%20%3Cspan%3E%5Cr%5Cn%20%20%20%20%20%20%20%20In%20order%20for%20us%20to%20evaluate%20and%20accept%20your%20PR%2C%20we%20ask%20that%20you%20sign%20a%20contribution%20license%20agreement.%20It%27s%20all%20electronic%20and%20will%20take%20just%20minutes.%20I%20promise%20there%27s%20no%20faxing.%20https%3A//cla.microsoft.com.%5Cr%5Cn%20%20%20%20%3C/span%3E%5Cr%5Cn%5Cr%5CnTTYL%2C%20MSBOT%3B%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222015-08-04T22%3A47%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/9287708%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/msftclas%22%7D%7D%2C%20%7B%22body%22%3A%20%22__%40korroz__%2C%20Thanks%20for%20signing%20the%20contribution%20license%20agreement%20so%20quickly%21%20Actual%20humans%20will%20now%20validate%20the%20agreement%20and%20then%20evaluate%20the%20PR.%5Cr%5Cn%3Cbr%20/%3EThanks%2C%20MSBOT%3B%22%2C%20%22created_at%22%3A%20%222015-08-04T23%3A07%3A12Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/9287708%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/msftclas%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks%20for%20your%20contribution%20%40korroz%20.%20I%20will%20merge%20this%20PR%20and%20then%20will%20make%20some%20changes%20to%20get%20it%20ready%20to%20merge%20it%20in%20KDev.%22%2C%20%22created_at%22%3A%20%222015-08-06T17%3A58%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/803796%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tushargupta51%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/208#issuecomment-127783346'>General Comment</a></b>
- <a href='https://github.com/msftclas'></a> Hi __@korroz__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
<span>
In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
</span>
TTYL, MSBOT;
- <a href='https://github.com/msftclas'></a> __@korroz__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;
- <a href='https://github.com/tushargupta51'></a> Thanks for your contribution @korroz . I will merge this PR and then will make some changes to get it ready to merge it in KDev.

<a href='https://www.codereviewhub.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/208?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/208?approve=1'></a>&nbsp;<a href='https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/208'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by brandwe** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/208'>Refresh</a>.<a href='#crh-update-none'></a>

**Comment by tushargupta51** [comment]
Thanks for your contribution @korroz . I will merge this PR and then will make some changes to get it ready to merge it in KDev.

**Comment by korroz** [comment]
Happy to help.

Note, however, that the default behavior is still unchanged. Which I think we agreed was to be changed to no claim transformation taking place. It would be a breaking change though.

**Comment by tushargupta51** [comment]
@korroz Yeah, you are right. We need to figure out how we are going to handle the defaults and the breaking change.

**Comment by brentschmaltz** [comment]
@korroz about the default. I know we all discussed mapping off by default, we have to be careful about a breaking change. There are some folks who are still concerned about removing the mapping by default so we broke this into two steps.

**Comment by korroz** [comment]
@brentschmaltz they are right to be concerned since it's a big break. Can't take breaking changes like this lightly.

At least with the instance level control of the claim mapping you don't have to disable/enable it application-wide. It was all or nothing before, or you could switch it on/off at runtime but then you'd have thread safety issues.

I suppose the best way to handle it would be something like changing the defaults in the next major version while retaining the old defaults in maintenance branches for older versions.

**Comment by brentschmaltz** [comment]
@korroz are you OK with leaving the default as is?

**Comment by korroz** [comment]
@brentschmaltz my opinion, like many others, are that the proper default should be that no mapping/filtering is performed. Having a non-magic default is the ideal. But I also recognize that, unfortunately, previous versions have had magic defaults already which users might depend on. So for pragmatic reasons I would keep the old default behavior until bumping major version. This depends a bit on if you're using [semantic versioning](http://semver.org/) or not. In the end, with instance level control of the mapping/filtering it is much easier to live with the magic defaults.

TL;DR I can live with it, especially with instance level control. But magic defaults are still wrong.

**Comment by brentschmaltz** [comment]
@korroz if I had a time machine ...

**Comment by korroz** [comment]
@brentschmaltz haha, indeed, if only. I know that feeling.
